### DISCUSSION TITLE
Collection::chunk doesn't allow to change the preserve_keys flag.

### DISCUSSION THREAD

**Discussion by erdemkeren** [laravel/framework#49483] [Ideas]
## Background: 

**PHP** `array_chunk` accepts 3 params: `array $array`, `int $length` and `bool $preserve_keys = false`.

Which means every chunk provided will be reindexed numerically starting from 0.

**Laravel** Collection::chunk uses a`array_chunk` under the hood. However it passes `true` as the value of `$preserve_keys` parameter (remember that PHP's default is false) and it doesn't allow us to pass an optional parameter.

Which means even if the chunk is different, the indexes will be preserved.

## Problem

Currently, I use the chunk keys as the index of the spreadsheet export keys.

Every chunk should start from the first column of the provided row.

For example:

```
A {1, 2, 3}
B {1, 2, 3}
...
```

with `$preserve_keys = true` what I get is:

```
A{1, 2, 3}
B{4, 5, 6}
...
```

Because I can't pass `$preserve_keys = false` to the method, I need to
1- Extend the collection and override the method
2- Use `array_chunk` 
or 3- Recalculate the index to match the default behaviour.

I could pass , `false` as the second paremeter instead.

So suggestion is:

```php
    public function chunk($size, $preserveKeys = true)
    {
        if ($size <= 0) {
            return new static;
        }

        $chunks = [];

        foreach (array_chunk($this->items, $size, $preserveKeys) as $chunk) {
            $chunks[] = new static($chunk);
        }

        return new static($chunks);
    }
```

## More

A similar isse was created here: https://github.com/laravel/framework/issues/29685. But it was redirected to the ideas and a PR was requested. Now that the repo is a public archive and we're redirected to here, I thought it is a good idea to discuss it again.

**Comment by AceXintense** [comment]
How has this not been implemented yet? I understand that making a change to keep preserveKeys inline with PHP will break things, but I believe the way @erdemkeren has implemented this allows users to pick wether to use it or not and will not break the framework for others due to it not overriding Laravel's current implementation.

I have just hit this issue myself and now I need to use the native php function to get around this quirk, If its too dangerous to do this directly in a function that could be used by millions then create a new function please. Having to go from collection to array is not ideal and is an unnecessary code smell.

I suspect this issue hasn't been reviewed due to the fact that its quite a niche use case but still!

**Reply by onlime** [reply]
> How has this not been implemented yet? I understand that making a change to keep preserveKeys inline with PHP will break things, but I believe the way @erdemkeren has implemented this allows users to pick wether to use it or not and will not break the framework for others due to it not overriding Laravel's current implementation.

same here. just wasted a lot of time 2 days ago, ended up implementing my own `Collection::chunkValues()` macro, as I couldn't use the current chunk() implementation anywhere.
In other use cases, I am using e.g. `Collection::forPage(...)->values()` but that's acceptable, as we're dealing with a single Collection.
